### DISCUSSION TITLE
JSON formatting?

### DISCUSSION THREAD

**Discussion by steveruizok** [supabase/supabase#1239] [Questions] [answered]
Hi, I'm having trouble with JSON in Supabase. For some reason I can't get values from a `jsonb` object until I open the JSON editor in Supabase, then click **Apply**.

![Kapture 2021-04-22 at 11 07 52](https://user-images.githubusercontent.com/23072548/115697041-1d11f680-a35b-11eb-862b-3e39524502e6.gif)

---

## What's not working

I'm using the JavaScript client to create rows in table named `projects`. The table includes a `jsonb` column named `document`. I am creating rows in the table using the following query:

```js
const document = JSON.stringify({ ... })

await supabase
    .from("projects")
    .insert([{ name, document }])
```

If I run a `SELECT * FROM projects` query, I can see the `document` column. Its value looks like this:

```
{"id":"1619036150909","name":"My Project","nodes":{},"globs":{},"groups":{},"pages":{"0":{"id":"0","name":"Page 1","type":0,"locked":false,"childIndex":0}},"version":"1","code":{"0":{"id":"0","childIndex":0,"name":"My Code","code":""}}}
```

However, if I then try to query into this JSON, I get no rows.

```sql
SELECT * from projects WHERE document->>'name' = 'My Project'
```

## A strange solution

**Here's where things get strange.** If I then go into the Table editor and open up the JSON editor panel and click **Apply**, then query will run correctly. See the GIF at the top for an example.

The resulting data appears to be identical except for alphabetical keys.

```
{"id":"1619036150909","code":{"0":{"id":"0","code":"","name":"My Code","childIndex":0}},"name":"My Project","globs":{},"nodes":{},"pages":{"0":{"id":"0","name":"Page 1","type":0,"locked":false,"childIndex":0}},"groups":{},"version":"1"}
```

## A solution or a bug?

From what I can tell (this discussion https://github.com/supabase/supabase/discussions/447#discussioncomment-251381) the solution is to call `insert` with an object, not a string of JSON data. 

```js
const document = { ... } // An object

await supabase
    .from("projects")
    .insert([{ name, document }])
```

This feels like a bug—I would expect the client to accept a string for `document`.

**Comment by fedeamura** [comment, marked as answer]
same bug here. any solution?

**Reply by steveruizok** [reply]
The solution is to pass an object, rather than a string. This means either passing it the original object or the parsed JSON.

**Comment by soedirgo** [comment]
Hmm, this is because a stringified JSON is a valid JSON string, which itself is valid JSON ([RFC7159](https://www.ietf.org/rfc/rfc7159.txt)). When you edit and save the jsonb value, the string gets `JSON.parse`'d, hence why it works.

On that note, if you add a row with a JSON string jsonb value like `"abc"`, saving/viewing the JSON value in the table editor will show an error.

I wonder if it's worth complying to the JSON spec (I vote yes), in which case we need to add `"` around JSON strings in the table editor. Thoughts @joshenlim?

**Reply by joshenlim** [reply]
gotcha on this - hmm, correct me if I'm wrong here but if the dashboard sends stringified JSONs when inserting rows for `jsonb` columns, we'll still not be able to get values from a jsonb object right? since the value in the column is a string and doing something like `SELECT * FROM table WHERE jsonb->>'key' = 'value'` will not return the expected rows?

**Reply by soedirgo** [reply]
Yeah, so with this change the "A strange solution" section above no longer works, because the stringified JSON is saved as a stringified JSON, not a `JSON.parse`'d stringified JSON. But I think this is the simplest way to disambiguate JSON strings, save from e.g. storing which rows are actually a string vs. a number, an array, or an object (FTR other types of JSON values work fine).

**Reply by soedirgo** [reply]
As of today, jsonb column values are shown properly but only when clicked:

https://user-images.githubusercontent.com/31685197/120254843-4eb4a080-c2bd-11eb-946b-9a1929d15781.mov
